src/core/net/ip6.cpp

```diff
@@ -603,27 +622,240 @@ otError Ip6::HandleOptions(Message &aMessage, Header &aHeader, bool &aForward)
     return error;
 }
 
-otError Ip6::HandleFragment(Message &aMessage)
+otError Ip6::HandleFragmentation(Message &aMessage, IpProto aIpProto)
+{
+    otError error = OT_ERROR_NONE;
+
+#if OPENTHREAD_ENABLE_IP6_FRAGMENTATION
+    Header         header;
+    FragmentHeader fragmentHeader;
+    Message *      fragment           = NULL;
+    uint16_t       fragmentCnt        = 0;
+    uint16_t       payloadFragment    = 0;
+    uint16_t       offset             = 0;
+    uint32_t       maxPayloadFragment = (1028 - aMessage.GetOffset() - sizeof(fragmentHeader)) & 0xfffffff8;
+    uint32_t       payloadLeft        = aMessage.GetLength() - aMessage.GetOffset();
+
+    if (aMessage.GetLength() > kMaxAssembledDatagramLength)
+    {
+        ExitNow(error = OT_ERROR_NO_BUFS);
+    }
+
+    VerifyOrExit(aMessage.Read(0, sizeof(header), &header) == sizeof(header), error = OT_ERROR_PARSE);
+    header.SetNextHeader(kProtoFragment);
+
+    fragmentHeader.Init();
+    fragmentHeader.SetIdentification(Random::NonCrypto::GetUint32());
+    fragmentHeader.SetNextHeader(aIpProto);
+    fragmentHeader.SetMoreFlag();
+
+    while (payloadLeft != 0)
+    {
+        if (payloadLeft < maxPayloadFragment)
+        {
+            fragmentHeader.ClearMoreFlag();
+            payloadFragment = payloadLeft;
+            payloadLeft     = 0;
+            otLogDebgIp6("Last Fragment");
+        }
+        else
+        {
+            payloadLeft -= maxPayloadFragment;
+            payloadFragment = maxPayloadFragment;
+        }
+        offset = fragmentCnt * (maxPayloadFragment >> 3);
+        fragmentHeader.SetOffset(offset);
+        VerifyOrExit((fragment = NewMessage(0)) != NULL, error = OT_ERROR_NO_BUFS);
+        SuccessOrExit(error = fragment->SetLength(aMessage.GetOffset() + sizeof(fragmentHeader) + payloadFragment));
+        header.SetPayloadLength(payloadFragment + sizeof(fragmentHeader));
+        VerifyOrExit(fragment->Write(0, sizeof(header), &header) == sizeof(header), error = OT_ERROR_NO_BUFS);
```

Comment:
A successful call to `SetLength()` above should mean that calls to `Write()` here and below will always succeed. It may be more appropriate to use `assert()` here and below in place of `VerifyOrExit()`.

Reply:
yes. changed to `assert()`.